core/textile/bucket/crypto/decrypter.go

```diff
@@ -0,0 +1,98 @@
+package crypto
+
+import (
+	"bufio"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"errors"
+	"io"
+	"os"
+
+	"github.com/odeke-em/go-utils/tmpfile"
+)
+
+const _16KB = 16 * 1024
+
+var DecryptErr = errors.New("message corrupt or incorrect keys")
+
+// NewDecryptReader creates an io.ReadCloser wrapping an io.Reader using the keys and iv
+// to decode the content using AES and verify HMAC.
+func NewDecryptReader(r io.Reader, aesKey, iv, hmacKey []byte) (io.ReadCloser, error) {
+	return newDecryptReader(r, aesKey, iv, hmacKey)
+}
+
+func newDecryptReader(r io.Reader, aesKey []byte, iv []byte, hmacKey []byte) (io.ReadCloser, error) {
+	mac := make([]byte, hmacSize)
+	h := hmac.New(hashFunc, hmacKey)
+	dst, err := tmpfile.New(&tmpfile.Context{
```

Comment:
Is there a way we can decrypt to memory instead of to disk?

Reply:
Maybe for big files this is not ideal though so does this at least store the encrypted files to disk and not the decrypted ones?

Reply:
Only encrypted content is written to the file... It is while it is being read that it gets decrypted.

Reply:
Is this file deleted after it finished reading?

Reply:
Yes it is. Once the Done() function is called on the `dst`. That is why that special tmpfile package is used.

Reply:
Nice